File: server/src/server/userFeeds.ml
open Request
open Response
open Util

let add_user_feed user_id source connection =
  (match%lwt Database.UserFeeds.create { user_id = user_id; feed_id = source } connection with
    | Ok _ ->
      Dream.log "[add_user_feed] source: %s - add success" source;
      Lwt.return_ok ()
    | Error e ->
      let message = Model.Error.Database.to_string e in
        Dream.log "[add_user_feed] source: %s - add failed with %s" source message;
        Lwt.return (Error (Model.Error.Database.to_frontend e)))

let get_all_user_feeds user_id connection =
  match%lwt Database.UserFeeds.feeds_by_user_id user_id connection with
    | Ok feeds ->
      Lwt.return feeds
    | Error e ->
      Dream.log "[get_all_user_feeds] - lookup failed with %s" (Model.Error.Database.to_string e);
      Lwt.return []

let remove_user_feed user_id source connection =
  match%lwt Database.UserFeeds.delete user_id source connection with
    | Ok _ ->
      Lwt.return_ok ()
    | Error e ->
      let message = Model.Error.Database.to_string e in
        Dream.log "[remove_user_feed] source: %s - add failed with %s" source message;
        Lwt.return (Error (Model.Error.Database.to_frontend e))

let get_items user_id connection =
  match%lwt Database.UserItems.all_items_by_user_id user_id connection with
    | Ok items ->
      Lwt.return items
    | Error e ->
      Dream.log "[get_items] user_id: %s - lookup failed with %s" user_id (Model.Error.Database.to_string e);
      Lwt.return []

let routes = [
  Dream.scope "/user_feeds" [Util.Middleware.cors; Util.Middleware.require_auth] [
    Dream.post "" (fun request ->
      let user_id = Dream.field request Util.Middleware.user_id |> Option.get in
      let%lwt body = Dream.body request in

      let req = body |> Yojson.Safe.from_string |> user_feed_add_request_of_yojson in
        match req with
          | Ok { uri } ->
            Dream.log "[/user_feeds POST] source: %s" uri;
            (match%lwt Dream.sql request (add_user_feed user_id uri) with
              | Ok _ ->
                Dream.log "[/user_feeds POST] source: %s - add success" uri;
                json { message = "ok" } status_response_to_yojson
              | Error e ->
                Dream.log "[/user_feeds POST] source: %s - add failed with %s" uri (Model.Error.Frontend.to_string e);
                throw_error e)
          | _ ->
            throw_error Model.Error.Frontend.BadRequest
    );

    Dream.get "/" (fun request ->
      let user_id = Dream.field request Util.Middleware.user_id |> Option.get in

      let _ = Dream.log "[/user_feeds/ GET]" in
      let%lwt feeds = Dream.sql request (get_all_user_feeds user_id) in
        json { feeds = List.map Model.Feed.Frontend.to_frontend feeds } all_feeds_response_to_yojson
    );

    Dream.get "/items" (fun request ->
      let user_id = Dream.field request Util.Middleware.user_id |> Option.get in

      let _ = Dream.log "[/user_feeds/ GET]" in
      let%lwt items = Dream.sql request (get_items user_id) in
        json { items = List.map Model.UserItem.Frontend.to_frontend items } items_response_to_yojson
    );

    Dream.delete "/:source" (fun request ->
      let user_id = Dream.field request Util.Middleware.user_id |> Option.get in
      let source = Dream.param request "source" in

      let _ = Dream.log "[/user_feeds/:source DELETE] source: %s" source in
      match%lwt Dream.sql request (remove_user_feed user_id source) with
        | Ok _ ->
          Dream.log "[/user_feeds/:source DELETE] source: %s - delete success" source;
          json { message = "ok" } status_response_to_yojson
        | Error e ->
          Dream.log "[/user_feeds/:source DELETE] source: %s - delete failed with %s" source (Model.Error.Frontend.to_string e);
          throw_error e
    );
  ]
]

File: ui/src/api/user-feeds.ts
import { request } from './util';
import { Feed } from '../models/feed';
import { Item, ItemEntity } from '../models/item';
import Users from './users';

interface AllFeedsResponse {
  feeds: Feed[];
}

interface ItemsResponse {
  items: ItemEntity[];
}

const UserFeeds = {
  async all(): Promise<Feed[]> {
    const response = await request(
      "/user_feeds/",
      {
        method: "GET",
        headers: {
          'Content-Type': 'application/json',
          'authentication': `Bearer ${Users.token()}`
        }
      }
    );

    if (response.ok) {
      const json: AllFeedsResponse = await response.json();
      return json.feeds
    } else {
      throw new Error(`UserFeeds.all - failed to fetch`);
    }
  },

  async items(): Promise<Item[]> {
    const response = await request(
      "/user_feeds/items",
      {
        method: "GET",
        headers: {
          'Content-Type': 'application/json',
          'authentication': `Bearer ${Users.token()}`
        }
      }
    );

    if (response.ok) {
      const json: ItemsResponse = await response.json();
      return json.items.map(({ item, metadata}) => ({ ...item, ...metadata }));
    } else {
      throw new Error(`UserFeeds.items - failed to fetch`);
    }
  },

  async recently_read_items(): Promise<Item[]> {
    const response = await request(
      "/user_feeds/items/read",
      {
        method: "GET",
        headers: {
          'Content-Type': 'application/json',
          'authentication': `Bearer ${Users.token()}`
        }
      }
    );

    if (response.ok) {
      const json: ItemsResponse = await response.json();
      return json.items.map(({ item, metadata}) => ({ ...item, ...metadata }));
    } else {
      throw new Error(`UserFeeds.items - failed to fetch`);
    }
  },

  async add(source: string): Promise<void> {
    const response = await request(
      "/user_feeds",
      {
        method: "POST",
        headers: {
          'Content-Type': 'application/json',
          'authentication': `Bearer ${Users.token()}`
        },
        body: JSON.stringify({ uri: source })
      }
    );

    if (!response.ok) {
      throw new Error(`UserFeeds.create(${source}) - failed to fetch`);
    }
  },

  async remove(id: string): Promise<void> {
    const response = await request(
      `/user_feeds/${encodeURIComponent(id)}`,
      {
        method: "DELETE",
        headers: {
          'Content-Type': 'application/json',
          'authentication': `Bearer ${Users.token()}`
        }
      }
    );

    if (!response.ok) {
      throw new Error(`UserFeeds.delete(${id}) - failed to fetch`);
    }
  }
}

export default UserFeeds;